book/src/proposals/abi-management.md

```diff
@@ -0,0 +1,434 @@
+# Solana ABI management process
+
+This document proposes Solana ABI management process. The ABI management process
+is an engineering practice and supporting technical framework to avoid
+introducing unintended incompatible ABI changes.
+
+# Problem
+
+The Solana ABI (binary interface to the cluster) is currently only defined 
+implicitly by the implementation, and requires a very careful eye to notice
+breaking changes. This makes it extremely difficult to upgrade the software
+on an existing cluster without rebooting the ledger.
+
+# Requirements and objectives
+
+- Unintended ABI changes can be detected as CI failures mechanically.
+- Newer impl must be able to process oldest data (since genesis) once we go
+  mainnet. (This is in stark contrast with regard to conventional software;
+  A decade-old SSL 3.0 doesn't work anymore, but the genesis block of Bitcoin
+  is still replayable today):
+  - Deserialization stability: Newer implementation (like system programs)
+    processes past object(transactions) exactly in the identical way when
+    replaying the ledger.
+  - Serialization stability: Newer implementation (like AccountsDB) produces
+    newer blobs exactly in the identical way unless its ABI is explicitly and
+    knowingly changed.
+- The objective of this proposal is to protect the ABI while sustaining rather
+  rapid development by opting for mechanical process rather than very long
+  human-driven auditing process. After all, while cluster instability must be
+  avoided, timely delivery of software improvements is desired for all
+  stakeholders of the Solana cluster.
+- Once cryptographically-signed, signed data blob must be identical, so no
+  in-place data format update is possible including inbound and outbound of the
+  online system. Also, considering the sheer volume of transactions we're aiming
+  to handle, retrospective in-place update is undesirable at best.
+
+# Solution
+
+Instead of natural human's eye due-diligence, which should be assumed to fail
+regularly, we need a systematic assurance of not breaking the cluster when
+changing the source code.
+
+For that purpose, we propose a two-fold mechanism of marking every ABI-related
+things in source code (`struct`, `enums`) with the new `#[solana_abi]` attribute
+and associating ABI-focussed unit tests with them.
+
+The attribute try to detect any unsanctioned changes to the marked ABI-related
+things when tests are run locally or on the CI, gathering closer attentions
+from engineers further on. However, the detection cannot be complete; no matter
+how hard we statically analyze the source code, it's still possible to break
+ABI.
+
+Thus, we confirm non-breakage by running paired tests for each ABI-related
+things which are separated for both serialization and deserialization and are
+versioned by the use of test fixtures containing actual binaries.
```

Comment:
The overall framing/scope of this design proposal is legit/sound? I think this is what we wanted from [the original comment](https://github.com/solana-labs/solana/pull/7415#discussion_r356392307), is this right?

Reply:
We may not actually need serialization and deserialization tests.   The solana_abi `digest=` field may be enough.  Once we have the digest computed, it should never change.

Reply:
I see! Thanks for clarification!